llvm/include/llvm-c/CAS/PluginAPI_functions.h

```
@@ -0,0 +1,257 @@
+ */
+LLCAS_PUBLIC unsigned llcas_digest_parse(llcas_cas_t,
+                                         const char *printed_digest,
+                                         uint8_t *bytes, size_t bytes_size,
+                                         char **error);
+
+/**
+ * Returns a string for the given digest bytes that can be passed to
+ * \c llcas_digest_parse.
+ *
+ * \param printed_id pointer to receive the printed digest string. The memory it
+ * points to needs to be released via \c llcas_string_dispose.
+ * \param error optional pointer to receive an error message if an error
+ * occurred. If set, the memory it points to needs to be released via
+ * \c llcas_string_dispose.
+ * \returns true if there was an error, false otherwise.
+ */
+LLCAS_PUBLIC bool llcas_digest_print(llcas_cas_t, llcas_digest_t,
+                                     char **printed_id, char **error);
+
+/**
+ * Provides the \c llcas_objectid_t value for the given \c llcas_digest_t.
+ *
+ * \param error optional pointer to receive an error message if an error
+ * occurred. If set, the memory it points to needs to be released via
+ * \c llcas_string_dispose.
+ * \returns true if there was an error, false otherwise.
+ */
+LLCAS_PUBLIC bool llcas_cas_get_objectid(llcas_cas_t, llcas_digest_t,
+                                         llcas_objectid_t *, char **error);
+
+/**
+ * \returns the \c llcas_digest_t value for the given \c llcas_objectid_t.
+ */
+LLCAS_PUBLIC llcas_digest_t llcas_objectid_get_digest(llcas_cas_t,
+                                                      llcas_objectid_t);
+
+/**
+ * Checks whether a \c llcas_objectid_t points to an existing object.
+ *
+ * \param error optional pointer to receive an error message if an error
+ * occurred. If set, the memory it points to needs to be released via
+ * \c llcas_string_dispose.
+ * \returns one of \c llcas_load_result_t.
+ */
+LLCAS_PUBLIC llcas_load_result_t llcas_cas_contains_object(llcas_cas_t,
```

Why is this `llcas_load_result_t` instead of `bool`? I assume you can't actually rely on this loading the object and you still need to separately load it

---

Allows flexibility for the plugin to return an error for whatever reason it wants. E.g. if it decides to do a network trip it may want to return an error for network failure.

Using `llcas_load_result_t` is only for having a "success/not-found/error" return value, without introducing yet another type, it doesn't necessarily imply that there is loading happening.
If it's confusing maybe I should rename `llcas_load_result_t` to something else, like `llcas_lookup_result_t`.

---

Ah that makes sense. I'm fine with either rename to lookup_result or tweak the doc comment to explain that it's not necessarily loading in this case.

---

Renamed to `llcas_lookup_result_t`.